File: anony/helpers/tv.py
import json
import os
import httpx
from pyrogram.types import InlineKeyboardMarkup, InlineKeyboardButton

# Load channels data
CHANNELS_FILE = os.path.join(os.path.dirname(__file__), "channels.json")

def load_channels():
    try:
        with open(CHANNELS_FILE, "r", encoding="utf-8") as f:
            return json.load(f)
    except Exception as e:
        print(f"Error loading channels.json: {e}")
        return []

def get_categories(channels=None):
    if channels is None:
        channels = load_channels()
    categories = list(set([c["category"] for c in channels if "category" in c]))
    categories.sort()
    return categories

def get_channels_by_category(category, channels=None):
    if channels is None:
        channels = load_channels()
    return [c for c in channels if c.get("category") == category]

def category_markup():
    categories = get_categories()
    keyboard = []
    row = []
    
    # Create 3-column rows
    for cat in categories:
        row.append(InlineKeyboardButton(cat, callback_data=f"tv_cat:{cat}"))
        if len(row) >= 3:
            keyboard.append(row)
            row = []
    
    if row:
        keyboard.append(row)
        
    keyboard.append([InlineKeyboardButton("❌ Close Play Menu", callback_data="help close")])
    return InlineKeyboardMarkup(keyboard)

def channel_markup(category, page=1):
    channels = get_channels_by_category(category)
    keyboard = []
    
    # 10 channels per page
    items_per_page = 10
    start_idx = (page - 1) * items_per_page
    end_idx = start_idx + items_per_page
    
    current_channels = channels[start_idx:end_idx]

    
    for ch in current_channels:
        # Two columns for channels
        if len(keyboard) > 0 and len(keyboard[-1]) < 2:
            keyboard[-1].append(InlineKeyboardButton(f"📺 {ch['title']}", callback_data=f"tv_ch:{ch['id']}"))
        else:
            keyboard.append([InlineKeyboardButton(f"📺 {ch['title']}", callback_data=f"tv_ch:{ch['id']}")])

            
    # Pagination
    nav_row = []
    if page > 1:
        nav_row.append(InlineKeyboardButton("⬅️ Back", callback_data=f"tv_page:{category}:{page-1}"))
    if end_idx < len(channels):
        nav_row.append(InlineKeyboardButton("Next ➡️", callback_data=f"tv_page:{category}:{page+1}"))
        
    if nav_row:
        keyboard.append(nav_row)
        
    # Back to categories
    keyboard.append([InlineKeyboardButton("🔙 Back to Categories", callback_data="tv_home")])
    return InlineKeyboardMarkup(keyboard)

async def fetch_stream_url(manifest_url):
    """Fetches the actual stream URL from the manifest endpoint."""
    from anony import config
    
    proxy_url = None
    if hasattr(config, "PROXY_URL") and config.PROXY_URL:
        proxy_url = config.PROXY_URL

    async with httpx.AsyncClient(proxy=proxy_url) as client:
        try:
            response = await client.get(manifest_url)
            response.raise_for_status()
            data = response.json()
            if data.get("status") == "ok" and "url" in data.get("data", {}):
                return data["data"]["url"]
            return None
        except Exception as e:
            print(f"Error fetching TV manifest: {e}")
            return None

File: internal/tv/tv.go
package tv

import (
	"encoding/json"
	"fmt"
	"net/http"
	"os"
	"sort"
	"time"

	"alexamusic/internal/bot"

	"github.com/amarnathcjd/gogram/telegram"
)

// Channel represents a TV station from channels.json.
type Channel struct {
	ID        string `json:"id"`
	Title     string `json:"title"`
	Category  string `json:"category"`
	Manifest  string `json:"manifest"`
	Thumbnail string `json:"thumbnail,omitempty"`
}

var channels []Channel

// Load loads channels from the JSON file.
func Load(path string) error {
	data, err := os.ReadFile(path)
	if err != nil {
		return err
	}
	return json.Unmarshal(data, &channels)
}

// GetCategories returns a sorted list of unique categories.
func GetCategories() []string {
	catMap := make(map[string]bool)
	for _, c := range channels {
		if c.Category != "" {
			catMap[c.Category] = true
		}
	}
	cats := make([]string, 0, len(catMap))
	for k := range catMap {
		cats = append(cats, k)
	}
	sort.Strings(cats)
	return cats
}

// GetByCategory returns channels for a specific category.
func GetByCategory(cat string) []Channel {
	var out []Channel
	for _, c := range channels {
		if c.Category == cat {
			out = append(out, c)
		}
	}
	return out
}

// GetByID returns a channel by its ID.
func GetByID(id string) *Channel {
	for _, c := range channels {
		if c.ID == id {
			return &c
		}
	}
	return nil
}

// CategoryMarkup builds the main TV category keyboard.
func CategoryMarkup() *telegram.ReplyInlineMarkup {
	cats := GetCategories()
	var buttons []telegram.KeyboardButton
	for _, cat := range cats {
		buttons = append(buttons, bot.InlineKeyboardButton(cat, "tv_cat:"+cat))
	}
	var rows [][]telegram.KeyboardButton
	for i := 0; i < len(buttons); i += 3 {
		end := i + 3
		if end > len(buttons) {
			end = len(buttons)
		}
		rows = append(rows, buttons[i:end])
	}
	rows = append(rows, []telegram.KeyboardButton{
		bot.InlineKeyboardButton("❌ Close Menu", "help close"),
	})
	return bot.Markup(rows...)
}

// ChannelMarkup builds a paged keyboard for a category.
func ChannelMarkup(category string, page int) *telegram.ReplyInlineMarkup {
	list := GetByCategory(category)
	perPage := 10
	start := (page - 1) * perPage
	end := start + perPage
	if end > len(list) {
		end = len(list)
	}

	var rows [][]telegram.KeyboardButton
	for i := start; i < end; i += 2 {
		row := []telegram.KeyboardButton{
			bot.InlineKeyboardButton("📺 "+list[i].Title, "tv_ch:"+list[i].ID),
		}
		if i+1 < end {
			row = append(row, bot.InlineKeyboardButton("📺 "+list[i+1].Title, "tv_ch:"+list[i+1].ID))
		}
		rows = append(rows, row)
	}

	// Navigation
	var nav []telegram.KeyboardButton
	if page > 1 {
		nav = append(nav, bot.InlineKeyboardButton("⬅️ Back", fmt.Sprintf("tv_page:%s:%d", category, page-1)))
	}
	if end < len(list) {
		nav = append(nav, bot.InlineKeyboardButton("Next ➡️", fmt.Sprintf("tv_page:%s:%d", category, page+1)))
	}
	if len(nav) > 0 {
		rows = append(rows, nav)
	}

	rows = append(rows, []telegram.KeyboardButton{
		bot.InlineKeyboardButton("🔙 Back to Categories", "tv_home"),
	})
	return bot.Markup(rows...)
}

// FetchStreamURL gets the real stream URL from the viu.lk manifest endpoint.
func FetchStreamURL(manifestURL string, proxyURL string) (string, error) {
	client := &http.Client{Timeout: 10 * time.Second}
	if proxyURL != "" {
		// handle proxy if needed, but for now simple direct
	}
	resp, err := client.Get(manifestURL)
	if err != nil {
		return "", err
	}
	defer resp.Body.Close()

	var data struct {
		Status string `json:"status"`
		Data   struct {
			URL string `json:"url"`
		} `json:"data"`
	}
	if err := json.NewDecoder(resp.Body).Decode(&data); err != nil {
		return "", err
	}
	if data.Status == "ok" && data.Data.URL != "" {
		return data.Data.URL, nil
	}
	return "", fmt.Errorf("manifest status not ok: %s", data.Status)
}
